src/programmatic_policy_learning/dsl/llm_primitives/llm_generator.py

```diff
@@ -0,0 +1,135 @@
+"""This module defines utilities for transforming the structured output of a
+LLM into a formal Grammar object.
+
+It processes nonterminals, terminals, and production rules to
+dynamically construct grammars for DSLs.
+"""
+
+import json
+import logging
+from typing import Any
+
+from prpl_llm_utils.models import PretrainedLargeModel
+from prpl_llm_utils.reprompting import query_with_reprompts
+from prpl_llm_utils.structs import Query
+
+from programmatic_policy_learning.dsl.generators.grammar_based_generator import Grammar
+from programmatic_policy_learning.dsl.llm_primitives.utils import (
+    JSONStructureRepromptCheck,
+)
+
+
+class LLMPrimitivesGenerator:
+    """A class to interact with an LLM, process its response, and generate a
+    Grammar object."""
+
+    def __init__(self, llm_client: PretrainedLargeModel | None):
+        """Initialize the generator with an LLM client.
+
+        Args:
+            llm_client (Any): An object or function to interact
+            with the LLM (e.g., OpenAI API client).
+        """
+        self.llm_client = llm_client
+
+    def query_llm(self, prompt: str) -> dict:
+        """Send a query to the LLM and return the parsed JSON response.
+
+        Args:
+            prompt (str): The query or prompt to send to the LLM.
+
+        Returns:
+            dict: The parsed JSON response from the LLM.
+        """
+        if self.llm_client is None:
+            raise ValueError("LLM client is not initialized.")
+
+        query = Query(prompt)
+        reprompt_checks = [JSONStructureRepromptCheck()]
+        response = query_with_reprompts(
+            self.llm_client,
+            query,
+            reprompt_checks,  # type: ignore[arg-type]
+            max_attempts=5,
+        )
+        logging.info("Response from LLM:")
+        logging.info(response)
+        return json.loads(response.text)
+
+    def create_grammar_from_response(
+        self, llm_output: dict[str, Any], object_types: list[Any]
```

Comment:
I don't know exactly how, but I think we should get away from this "object_types" thing as much as we can, because it's going to be an obstacle as soon as we try to run these approaches in a new set of environments where "object_types" is no longer a thing